Add Blas compaction for raytracing

msat-huawei opened the pull request:
## What does this PR do?

Add raytracing acceleration compaction for Blas instances in the `RayTracingFeatureProcessor` as described [here](https://developer.nvidia.com/blog/tips-acceleration-structure-compaction/). This can save a lot of memory in large scenes. Usually the size of a Blas is about halfed after compaction.

We refactored the `RayTracingFeatureProcessor` so we can keep track of the state of the Blas (Blas Instance Created -> Blas Created -> Blas compaction query inserted -> Blas compaction enqueued -> Uncompacted Blas deleted). These states are defined implicitly through the lists the Blas instance is part of. This was necessary because the compaction happens over multiple frames.
The compacted sizes of Blas are queried using the newly introduced classes `RayTracingCompactionQueryPool` and `RayTracingCompactionQueryPool`
We also added the ability to limit the number of meshes for which Blas instances are created each frame. When adding all Blas instances in the same frame and then compact the Blas in the same frame, we get a memory spike while the uncompacted and compacted Blas are present at the same time. Limiting the number of Blas that are created each frame reduces this memory spike. This is turned off by default, but can be enabled through a registry setting.

This PR also fixes a MultGPU bug, where the `BeginFrame` function was called by each `RayTracingAccelerationStructurePass`, so for each device.

## How was this PR tested?

Tested on Windows with DX12 and Vulkan with different levels

martinwinter-huawei commented on Gems/Atom/RPI/Code/Include/Atom/RPI.Reflect/RPISystemDescriptor.h:
```diff
@@ -26,6 +26,24 @@ namespace AZ
+            bool m_enableBlasCompaction = true;
+
+            //! The maximum number of meshes for which Blas instances are created each frame
+            //! Can be used to limit peak memory consumption for raytracing when Blas compaction is enabled
```
What about `INT32_MAX` (that would simplify the usage which otherwise checks against > 0 first) as default parameter?

msat-huawei replied:
The value can be set in a setreg file. Setting it to `-1` there is much easier than setting it to `INT32_MAX`

msat-huawei commented on Gems/Atom/Feature/Common/Code/Source/RayTracing/RayTracingFeatureProcessor.cpp:
```diff
@@ -165,7 +178,9 @@ namespace AZ
             MeshBlasInstance meshBlasInstance;
             meshBlasInstance.m_count = 1;
-            meshBlasInstance.m_subMeshes.push_back(SubMeshBlasInstance{ rayTracingBlas });
+            SubMeshBlasInstance subMeshBlasInstance;
```
The struct now contains two Blas. I now explicitly set the uncompacted Blas so we don't get some weird runtime error if someone would change the order in the struct. I did add a move into the vector though now.

msat-huawei commented on Gems/Atom/Feature/Common/Code/Source/RayTracing/RayTracingFeatureProcessor.cpp:
```diff
@@ -842,6 +852,228 @@ namespace AZ
+                                                          ->GetResult();
+                                        compactedBlasToCreate[assetId][subMeshIdx][deviceId] = result;
+                                        {
+                                            // TODO debug delete
```
This was some left over debug code -> removed

msat-huawei commented on Gems/Atom/Feature/Common/Code/Source/RayTracing/RayTracingFeatureProcessor.cpp:
```diff
@@ -842,6 +852,228 @@ namespace AZ
+                                });
+                        }
+                    }
+                }
```
I refactored my code a bit and now got rid of the second for loop

moudgils commented on Gems/Atom/RHI/Code/Include/Atom/RHI/DeviceRayTracingCompactionQueryPool.h:
```diff
@@ -0,0 +1,65 @@
+
+namespace AZ::RHI
+{
+    class DeviceRayTracingCompactionQueryPool;
```
Require comments on new classes + API

moudgils commented on Gems/Atom/RHI/Code/Include/Atom/RHI/RayTracingAccelerationStructure.h:
```diff
@@ -104,6 +104,11 @@ namespace AZ::RHI
             const RayTracingBlasDescriptor* descriptor,
             const RayTracingBufferPools& rayTracingBufferPools);
 
+        ResultCode CreateCompactedBuffers(
```
Comments on new API

moudgils commented on Gems/Atom/RHI/Code/Include/Atom/RHI/RayTracingCompactionQueryPool.h:
```diff
@@ -0,0 +1,45 @@
+{
+    class RayTracingCompactionQueryPool;
+
+    class RayTracingCompactionQuery : public MultiDeviceObject
```
Documentation for new classes + API

msat-huawei replied:
Added comments

moudgils commented on Gems/Atom/RHI/DX12/Code/Source/RHI/RayTracingCompactionQueryPool.h:
```diff
@@ -0,0 +1,88 @@
+                AZStd::vector<AZStd::pair<RHI::Ptr<RayTracingCompactionQuery>, int>> enqueuedQueries;
+            };
+
+            AZStd::array<Buffers, RHI::Limits::Device::FrameCountMax + 1> m_buffers;
```
Why the + 1? Maybe add that as a comment

msat-huawei replied:
The +1 is not needed here at all -> Removed

moudgils replied:
While you are at it see if you can remove other instances of +1 used in this PR.

msat-huawei replied:
I removed the `+1` from some more places:
- Vulkan query pool
- Enqueuing compactions for Blas
- Enqueuing deletions of the uncompacted Blas

moudgils commented on Gems/Atom/RHI/DX12/Code/Source/RHI/RayTracingCompactionQueryPool.h:
```diff
@@ -0,0 +1,88 @@
+            void BeginFrame(int frame) override;
+            //////////////////////////////////////////////////////////////////////////
+
+            struct Buffers
```
Recommend: Buffers -> QueryPoolBuffers.

moudgils commented on Gems/Atom/RHI/DX12/Code/Source/RHI/RayTracingCompactionQueryPool.h:
```diff
@@ -0,0 +1,88 @@
+                AZStd::vector<AZStd::pair<RHI::Ptr<RayTracingCompactionQuery>, int>> enqueuedQueries;
+            };
+
+            AZStd::array<Buffers, RHI::Limits::Device::FrameCountMax + 1> m_buffers;
```
Recommendation: m_buffers->m_queryPoolBuffers

moudgils commented on Gems/Atom/RHI/Metal/Code/Source/RHI/CommandList.h:
```diff
@@ -69,8 +69,16 @@ namespace AZ
-            void BuildBottomLevelAccelerationStructure(const RHI::DeviceRayTracingBlas& rayTracingBlas) override;
+            void BuildBottomLevelAccelerationStructure(const RHI::DeviceRayTracingBlas &rayTracingBlas) override;
             void UpdateBottomLevelAccelerationStructure(const RHI::DeviceRayTracingBlas& rayTracingBlas) override;
+            void QueryBlasCompactionSizes(
```
Recommendation: One line per param.

msat-huawei commented on Gems/Atom/Feature/Common/Code/Source/RayTracing/RayTracingAccelerationStructurePass.cpp:
```diff
@@ -218,53 +219,122 @@ namespace AZ
+                        AZ_Assert(!enqueuedForCompaction, "All or none Blas of an asset need to be compacted");
+                    }
+                }
+                AZStd::lock_guard lock(rayTracingFeatureProcessor->GetBlasBuiltMutex());
```
Added a scope around the lock and the code that needs the lock

msat-huawei commented on Gems/Atom/Feature/Common/Code/Source/RayTracing/RayTracingAccelerationStructurePass.cpp:
```diff
@@ -218,53 +219,122 @@ namespace AZ
+                    }
+                    changedBlasList.push_back(submeshBlasInstance.m_blas->GetDeviceRayTracingBlas(context.GetDeviceIndex()).get());
+                }
+                AZStd::lock_guard lock(rayTracingFeatureProcessor->GetBlasBuiltMutex());
```
Added a scope around the lock and the code that needs the lock

msat-huawei commented on Gems/Atom/Feature/Common/Code/Source/RayTracing/RayTracingAccelerationStructurePass.cpp:
```diff
@@ -218,53 +219,122 @@ namespace AZ
+                    auto query = submeshBlasInstance.m_compactionSizeQuery;
+                    if (query)
+                    {
+                        context.GetCommandList()->CompactBottomLevelAccelerationStructure(
```
The compaction happens once for every Blas, a few frames after it was built

moudgils commented on Gems/Atom/RHI/DX12/Code/Source/RHI/RayTracingCompactionQueryPool.cpp:
```diff
@@ -0,0 +1,141 @@
+                }
+            }
+
+            m_currentBufferIndex = (m_currentFrame + 1) % m_queryPoolBuffers.size();
```
Do we need still need  '+ 1' here

msat-huawei replied:
The `+1` is needed, but instead of `m_currentFrame ` it should have been `m_currentBufferIndex`. `m_currentFrame` still worked as its also incremented every frame

moudgils commented on Gems/Atom/RHI/Vulkan/Code/Source/RHI/RayTracingCompactionQueryPool.cpp:
```diff
@@ -0,0 +1,114 @@
+
+        RHI::ResultCode RayTracingCompactionQueryPool::InitInternal(RHI::RayTracingCompactionQueryPoolDescriptor desc)
+        {
+            int queryPoolSize = desc.m_budget * (RHI::Limits::Device::FrameCountMax + 1);
```
Do we need still need  '+ 1' here

msat-huawei replied:
No, removed

msat-huawei commented:
In addition to the propsed fixes I also fixed a validation warning introduced by me here. In Vulkan, the queries need to be reset before the first use, and after each use.

msat-huawei commented:
After trying the [startergame-assets](https://github.com/o3de/startergame-assets) project I found two new problems in this PR. One of them is not solved yet:
1. The check for the budget of the compaction query pool was at the wrong place. I moved it to the right one in the last commit.
2. There is a problem with the Terrain system. The `RayTracingFeatureProcessor` update call was previously called by the `RayTracingAccelerationStructurePass`. I moved this to the `Render` function of the `FeatureProcessor` because multiple acceleration structure passes exist when multiple devices are present. Unfortunately this interferes with the Terrain system which add/removes meshes in the raytracing scene in its `Render` function. I don't have a solution for this problem yet.

msat-huawei commented on Gems/Atom/Feature/Common/Code/Include/Atom/Feature/RayTracing/RayTracingFeatureProcessorInterface.h:
```diff
@@ -341,17 +343,28 @@ namespace AZ::Render
         //! Retrieves the GPU buffer containing information for all ray tracing materials.
         virtual const Data::Instance<RPI::Buffer> GetMaterialInfoGpuBuffer() const = 0;
 
-        //! If necessary recreates TLAS buffers and updates the ray tracing SRGs. Should only be called by the
-        //! RayTracingAccelerationStructurePass. Returns the current revision.
-        virtual uint32_t BeginFrame() = 0;
-
```
This function should only ever be called from the acceleration structure passes. So it does not make sense to have this in the interface.
I did remove this function as a whole at one point and did not add the return value when I added it back. I don't think its obvious what is returned here. I would expect a frame number not a raytracing revision.

jhmueller-huawei replied:
The same could be said about other methods in the interface, like e.g. `GetRevision()` as well. It's just a bit inconsistent.

msat-huawei replied:
I readded the `BeginFrame` function to the interface now

jhmueller-huawei commented on Gems/Atom/Feature/Common/Code/Source/RayTracing/RayTracingFeatureProcessor.cpp:
```diff
@@ -842,6 +864,216 @@ namespace AZ
+
+            if (changed)
+            {
+                m_revision++;
```
That's why it's a nit.
